### PR TITLE
[libcurl/xxx] Conan v2 migration

### PR DESCRIPTION
**Pull request by MartinDelille**
Specify library name and version:  **libcurl/xxx**

Replacing #12377 after merge conflict.
I rebased on master and will try to reference all the ongoing issues.

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/libcurl/all/conanfile.py [hunk 1]
```diff
@@ -472,150 +506,89 @@ def _arm_version(self, arch):
...
-            self.run("{} -fiv".format(tools.get_env("AUTORECONF") or "autoreconf"), win_bash=tools.os_info.is_windows, run_environment=True)
-
-            # fix generated autotools files to have relocatable binaries
-            if is_apple_os(self):
-                replace_in_file(self, "configure", "-install_name \\$rpath/", "-install_name @rpath/")
-
-            self.run("chmod +x configure")
-
-            # run configure with *LD_LIBRARY_PATH env vars it allows to pick up shared openssl
-            with tools.run_environment(self):
```
Comment (MartinDelille) [inline recipes/libcurl/all/conanfile.py:487]: It seems there is still an issue here. See https://github.com/conan-io/conan-center-index/pull/12377#discussion_r962293843

File: recipes/libcurl/all/conanfile.py [hunk 2]
```diff
@@ -231,14 +235,19 @@ def imports(self):
         #     but does not work on OS X 10.11 with SIP)
         # 2. copying dylib's to the build directory (fortunately works on OS X)
         if self.settings.os == "Macos":
-            self.copy("*.dylib*", dst=self._source_subfolder, keep_path=False)
+            copy(self, "*.dylib*", src=self.build_folder, dst=self.source_folder, keep_path=False)
```
Comment (MartinDelille) [inline recipes/libcurl/all/conanfile.py:238]: In continuation of https://github.com/conan-io/conan-center-index/pull/12377#discussion_r963088772

File: recipes/libcurl/all/test_package/conanfile.py [hunk 3]
```diff
@@ -25,14 +21,11 @@ def build(self):
 
     @property
     def _test_executable(self):
-        if self.settings.os == "Windows":
-            return os.path.join("bin", "test_package.exe")
-        else:
-            return os.path.join("bin", "test_package")
+        return os.path.join(self.cpp.build.bindirs[0], "test_package")
```
Comment (MartinDelille) [inline recipes/libcurl/all/test_package/conanfile.py:24]: I simplified this after @SpaceIm remark here: https://github.com/conan-io/conan-center-index/pull/12377#pullrequestreview-1096719932

Unfortunately I now have a error [here](https://c3i.jfrog.io/c3i/misc/logs/pr/12956/1-linux-gcc/libcurl/7.77.0//66fe4987fe114c052c300157fb662fdbd5cea929-test.txt):

```
ERROR: libcurl/7.77.0 (test package): Error in test() method, line 22
	self.run(self._test_executable, run_environment=True)
while calling '_test_executable', line 18
	return os.path.join(self.cpp.build.bindirs[0], "test_package")
	TypeError: 'NoneType' object is not subscriptable
```

File: recipes/libcurl/all/conanfile.py [hunk 4]
```diff
@@ -213,13 +206,27 @@ def build_requirements(self):
...
+    def generate(self):
+        if self._is_using_cmake_build:
+            self._generate_with_cmake()
+        else:
+            self._generate_with_autotools()
+        ms = VirtualBuildEnv(self)
+        ms.generate()
+        if not cross_building(self):
+            env = VirtualRunEnv(self)
+            env.generate(scope="build")
```
Comment (MartinDelille) [inline recipes/libcurl/all/conanfile.py:229]: from https://github.com/conan-io/conan-center-index/pull/12850#pullrequestreview-1108879227

**Comment by prince-chrismc** [comment]
#12888 🚀

File: recipes/libcurl/all/conanfile.py [hunk 5]
```diff
@@ -213,13 +206,27 @@ def build_requirements(self):
         else:
             self.tool_requires("libtool/2.4.6")
             self.tool_requires("pkgconf/1.7.4")
-            if self._settings_build.os == "Windows" and not tools.get_env("CONAN_BASH_PATH"):
+            if self._settings_build.os == "Windows" and not get_env("CONAN_BASH_PATH") and not self.conf.get("tools.microsoft.bash:path", default=False, check_type=bool):
                 self.tool_requires("msys2/cci.latest")
 
+    def layout(self):
+        cmake_layout(self, src_folder="src")
```
Comment (prince-chrismc) [inline recipes/libcurl/all/conanfile.py:213]: It's autotools no?

File: recipes/libcurl/all/test_package/conanfile.py (hunk 3 above)
Comment (prince-chrismc) [reply recipes/libcurl/all/test_package/conanfile.py:24]: Dunno if it's related but it's missing the test type and sf reference 

#12935 should be a decent example

File: recipes/libcurl/all/conanfile.py [hunk 6]
```diff
@@ -213,13 +206,28 @@ def build_requirements(self):
         else:
             self.tool_requires("libtool/2.4.6")
             self.tool_requires("pkgconf/1.7.4")
-            if self._settings_build.os == "Windows" and not tools.get_env("CONAN_BASH_PATH"):
+            if self._settings_build.os == "Windows" and not get_env("CONAN_BASH_PATH") and not self.conf.get("tools.microsoft.bash:path", default=False, check_type=bool):
                 self.tool_requires("msys2/cci.latest")
 
+    def layout(self):
+        if self._is_using_cmake_build:
```
Comment (prince-chrismc) [inline recipes/libcurl/all/conanfile.py:213]: You can not conditionally set the layout.

This sets up important things on the package info... which AFAIK could explain why the test package does not work

Comment (SpaceIm) [reply recipes/libcurl/all/conanfile.py:213]: Use `basic_layout` if not self._is_using_cmake_build

Comment (prince-chrismc) [reply recipes/libcurl/all/conanfile.py:213]: New advice 😱

Comment (SpaceIm) [reply recipes/libcurl/all/conanfile.py:213]: Important: define a consistent src folder if there are conditional layouts.

Comment (uilianries) [reply recipes/libcurl/all/conanfile.py:213]: I asked about conditional layout here: https://github.com/conan-io/conan/issues/11876

File: recipes/libcurl/all/conanfile.py [hunk 7]
```diff
@@ -1,15 +1,18 @@
...
+from conan.tools.cmake import CMake, CMakeToolchain, cmake_layout
+from conan.tools.env import VirtualBuildEnv, VirtualRunEnv
+from conan.tools.files import apply_conandata_patches, copy, download, get, load, replace_in_file, rm, rmdir, save
+from conan.tools.gnu import Autotools, AutotoolsToolchain, AutotoolsDeps, PkgConfigDeps
+from conan.tools.microsoft import is_msvc, unix_path
 from conan.tools.scm import Version
-from conan.tools.files import replace_in_file, rmdir, chdir, get, download, save, load, apply_conandata_patches
-from conan.tools.apple import is_apple_os
+# TODO: Migrate
+from conans.tools import get_env
```
Comment (prince-chrismc) [inline recipes/libcurl/all/conanfile.py:12]: This has been dropped in other PRs... not sure thats the right thing or not

File: recipes/libcurl/all/conanfile.py [hunk 8]
```diff
@@ -213,13 +206,28 @@ def build_requirements(self):
         else:
             self.tool_requires("libtool/2.4.6")
             self.tool_requires("pkgconf/1.7.4")
-            if self._settings_build.os == "Windows" and not tools.get_env("CONAN_BASH_PATH"):
+            if self._settings_build.os == "Windows" and not get_env("CONAN_BASH_PATH") and not self.conf.get("tools.microsoft.bash:path", default=False, check_type=bool):
                 self.tool_requires("msys2/cci.latest")
 
+    def layout(self):
+        if self._is_using_cmake_build:
+            cmake_layout(self, src_folder="src")
```
Comment (prince-chrismc) [inline recipes/libcurl/all/conanfile.py:214]:
```suggestion
        if self._is_using_cmake_build:
            cmake_layout(self, src_folder="src")
        else:
            basic_layout(self, src_folder="src")
```

File: recipes/libcurl/all/conanfile.py [hunk 9]
```diff
@@ -1,15 +1,18 @@
...
 from conan.errors import ConanInvalidConfiguration
-from conan.tools.microsoft import is_msvc
+from conan.tools.apple import is_apple_os, fix_apple_shared_install_name
 from conan.tools.build import cross_building
+from conan.tools.cmake import CMake, CMakeToolchain, cmake_layout
+from conan.tools.env import VirtualBuildEnv, VirtualRunEnv
+from conan.tools.files import apply_conandata_patches, copy, download, get, load, replace_in_file, rm, rmdir, save
+from conan.tools.gnu import Autotools, AutotoolsToolchain, AutotoolsDeps, PkgConfigDeps
+from conan.tools.microsoft import is_msvc, unix_path
 from conan.tools.scm import Version
```
Comment (prince-chrismc) [inline recipes/libcurl/all/conanfile.py:10]:
```suggestion
from conan.tools.scm import Version
from conan.tools.layouts import basic_layout
```

Comment (SpaceIm) [reply recipes/libcurl/all/conanfile.py:10]: it's `conan.tools.layout`

Comment (prince-chrismc) [reply recipes/libcurl/all/conanfile.py:10]: fixed typo

File: recipes/libcurl/all/test_package/conanfile.py (hunk 3 above)
Comment (MartinDelille) [reply recipes/libcurl/all/test_package/conanfile.py:27]: It doesn't help. I can't even reproduce it locally using the following docker image: `conanio/gcc5`

**Comment by MartinDelille** [comment]
@czoido

File: recipes/libcurl/all/conanfile.py [hunk 10]
```diff
@@ -213,13 +210,30 @@ def build_requirements(self):
...
+            cmake_layout(self, src_folder="src")
+        else:
+            basic_layout(self, src_folder="src")
+
     def source(self):
         get(self, **self.conan_data["sources"][self.version],
-                  destination=self._source_subfolder, strip_root=True)
-        download(self, "https://curl.haxx.se/ca/cacert.pem", "cacert.pem", verify=True)
+                  destination=self.source_folder, strip_root=True)
+        download(self, "https://curl.haxx.se/ca/cacert.pem", "cacert.pem", verify=True, sha256="2cff03f9efdaf52626bd1b451d700605dc1ea000c5da56bd0fc59f8f43071040")
```
Comment (MartinDelille) [inline recipes/libcurl/all/conanfile.py:225]: I just had to update the sha256 that changed. Shouldn't we remove it to prevent build failure ?

**Comment by MartinDelille** [comment]
Any idea why the test is failing ?

```
ERROR: libcurl/7.84.0 (test package): Error in test() method, line 22
	self.run(self._test_executable, run_environment=True)
while calling '_test_executable', line 18
	return os.path.join(self.cpp.build.bindirs[0], "test_package")
	TypeError: 'NoneType' object is not subscriptable
```

I don't reproduce locally.

File: recipes/libcurl/all/test_v1_package/conanfile.py [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+    generators = "cmake", "cmake_find_package"
+
+    def build(self):
+        cmake = CMake(self)
+        cmake.configure()
+        cmake.build()
+
+    @property
+    def _test_executable(self):
+        return os.path.join(self.cpp.build.bindirs[0], "test_package")
```
[the remark this answers is not in the document]
Comment (MartinDelille) [reply recipes/libcurl/all/test_v1_package/conanfile.py:18]: I tried here so let's see: 3d5aec88c

**Comment by MartinDelille** [comment]
This is really weird: it looks like my latest change wasn't taken into account.

**Comment by prince-chrismc** [comment]
You changed the wrong file 🙈 there's two test packages. The v1 needs the old run.

Thats totally confusing

**Review by prince-chrismc — APPROVED** [review]

File: recipes/libcurl/all/conanfile.py (hunk 7 above)
Comment (MartinDelille) [reply recipes/libcurl/all/conanfile.py]: Indeed: here is an example 33c236a4a8c6d0fd95f2c27dfda13db7f2474052

I dropped it here: 9f24973c0

File: recipes/libcurl/all/conanfile.py [hunk 12]
```diff
@@ -1,15 +1,17 @@
...
-import functools
-from conans import ConanFile, AutoToolsBuildEnvironment, CMake, tools
+from conan import ConanFile
 from conan.errors import ConanInvalidConfiguration
-from conan.tools.microsoft import is_msvc
+from conan.tools.apple import is_apple_os, fix_apple_shared_install_name
 from conan.tools.build import cross_building
+from conan.tools.cmake import CMake, CMakeToolchain, cmake_layout
+from conan.tools.env import VirtualBuildEnv, VirtualRunEnv
+from conan.tools.files import apply_conandata_patches, copy, download, get, load, replace_in_file, rm, rmdir, save
```
Comment (jwillikers) [inline recipes/libcurl/all/conanfile.py:7]:
```suggestion
from conan.tools.files import apply_conandata_patches, copy, download, export_conandata_patches, get, load, replace_in_file, rm, rmdir, save
```

File: recipes/libcurl/all/conanfile.py (hunk 10 above)
Comment (jwillikers) [reply recipes/libcurl/all/conanfile.py:223]: I don't think that would be safe. Is there a way to reference a download that won't be updated?

File: recipes/libcurl/all/conanfile.py [hunk 13]
```diff
@@ -213,13 +208,30 @@ def build_requirements(self):
         else:
             self.tool_requires("libtool/2.4.6")
             self.tool_requires("pkgconf/1.7.4")
-            if self._settings_build.os == "Windows" and not tools.get_env("CONAN_BASH_PATH"):
+            if self._settings_build.os == "Windows" and not self.conf.get("tools.microsoft.bash:path", default=False, check_type=bool):
```
Comment (jwillikers) [inline recipes/libcurl/all/conanfile.py:211]: Should `win_bash` be set to `True` when building on Windows?

**Review by jwillikers — CHANGES_REQUESTED** [review]

**Review by prince-chrismc — APPROVED** [review]

File: recipes/libcurl/all/conanfile.py [hunk 14]
```diff
@@ -195,6 +187,9 @@ def requirements(self):
         if self.options.with_c_ares:
             self.requires("c-ares/1.18.1")
 
+    def package_id(self):
+        del self.info.settings.compiler
+
     def validate(self):
         if self.options.with_ssl == "schannel" and self.settings.os != "Windows":
```
[the remark this answers is not in the document]
Comment (uilianries) [reply recipes/libcurl/all/conanfile.py:193]: Excellent observation

File: recipes/libcurl/all/conanfile.py [hunk 15]
```diff
@@ -213,13 +207,30 @@ def build_requirements(self):
         else:
             self.tool_requires("libtool/2.4.6")
             self.tool_requires("pkgconf/1.7.4")
-            if self._settings_build.os == "Windows" and not tools.get_env("CONAN_BASH_PATH"):
+            if self._settings_build.os == "Windows" and not self.conf.get("tools.microsoft.bash:path", default=False, check_type=bool):
```
Comment (uilianries) [inline recipes/libcurl/all/conanfile.py:210]:
```suggestion
            if self._settings_build.os == "Windows" and not self.conf.get("tools.microsoft.bash:path", default=False, check_type=str):
```
A path is a string, not a boolean. I know, there are other recipes using it, but they are incorrect too.

File: recipes/libcurl/all/conanfile.py (hunk 10 above)
Comment (uilianries) [reply recipes/libcurl/all/conanfile.py:222]: Please, take a look on ApprovalTest conandata.yml as reference to download that license file: https://github.com/conan-io/conan-center-index/blob/master/recipes/approvaltests.cpp/all/conandata.yml

**Review by uilianries — COMMENTED** [review]
Looking good, just need some few updates.

File: recipes/libcurl/all/conanfile.py (hunk 10 above)
Comment (MartinDelille) [reply recipes/libcurl/all/conanfile.py:222]: The ApprovalTest example doesn't has to mix between tar.gz and pem files.

File: recipes/libcurl/all/conanfile.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (MartinDelille) [reply recipes/libcurl/all/conanfile.py]: Done here: 63c83a786

Can you explain me why ?

File: recipes/libcurl/all/conanfile.py (hunk 13 above)
Comment (MartinDelille) [reply recipes/libcurl/all/conanfile.py]: Is it enough ? 8c616b314

File: recipes/libcurl/all/conanfile.py [hunk 16]
```diff
@@ -213,13 +207,30 @@ def build_requirements(self):
         else:
             self.tool_requires("libtool/2.4.6")
             self.tool_requires("pkgconf/1.7.4")
-            if self._settings_build.os == "Windows" and not tools.get_env("CONAN_BASH_PATH"):
+            if self._settings_build.os == "Windows" and not self.conf.get("tools.microsoft.bash:path", default=False, check_type=str):
                 self.tool_requires("msys2/cci.latest")
```
Comment (jwillikers) [inline recipes/libcurl/all/conanfile.py:211]: I think like this:

```suggestion
                if self._settings_build.os == "Windows":
                    win_bash = True
                    if not self.conf.get("tools.microsoft.bash:path", default=False, check_type=str):
                        self.tool_requires("msys2/cci.latest")
```

File: recipes/libcurl/all/conanfile.py [hunk 17]
```diff
@@ -195,15 +186,18 @@ def requirements(self):
...
-        if self.options.with_ssl == "darwinssl" and not is_apple_os(self):
+        if self.info.options.with_ssl == "darwinssl" and not is_apple_os(self):
             raise ConanInvalidConfiguration("darwinssl only suppported on Apple like OS (Macos, iOS, watchOS or tvOS).")
-        if self.options.with_ssl == "wolfssl" and self._is_using_cmake_build and Version(self.version) < "7.70.0":
+        if self.info.options.with_ssl == "wolfssl" and self._is_using_cmake_build and Version(self.version) < "7.70.0":
             raise ConanInvalidConfiguration("Before 7.70.0, libcurl has no wolfssl support for Visual Studio or \"Windows to Android cross compilation\"")
-        if self.options.with_ssl == "openssl":
-            if self.options.with_ntlm and self.options["openssl"].no_des:
+        if self.info.options.with_ssl == "openssl":
+            if self.info.options.with_ntlm and self.info.options["openssl"].no_des:
```
Comment (uilianries) [inline recipes/libcurl/all/conanfile.py:200]:
```suggestion
            openssl = self.dependencies["openssl"]
            if self.info.options.with_ntlm and openssl.options.no_des:
```

Ops! https://docs.conan.io/en/latest/reference/conanfile/dependencies.html#dependencies-interface

**Review by jwillikers — DISMISSED** [review]

File: recipes/libcurl/all/conanfile.py [hunk 18]
```diff
@@ -213,13 +208,32 @@ def build_requirements(self):
         else:
             self.tool_requires("libtool/2.4.6")
             self.tool_requires("pkgconf/1.7.4")
```
Comment (uilianries) [inline recipes/libcurl/all/conanfile.py:210]:
```suggestion
            if not self.conf.get("tools.gnu:pkg_config", default=False, check_type=str)
                self.tool_requires("pkgconf/1.7.4")
```

**Review by jwillikers — DISMISSED** [review]

**Review by jwillikers — APPROVED** [review]

**Review by prince-chrismc — APPROVED** [review]

**Comment by prince-chrismc** [comment]
Oh yes amazing~!!!!!! merge!!!!! bot now please!!!!!

**Comment by MartinDelille** [comment]
Yes that last for a long time!

Unfortunately there was still two issues I addressed here:
- https://github.com/conan-io/conan-center-index/pull/12956#discussion_r971814925
- https://github.com/conan-io/conan-center-index/pull/12956#discussion_r971816199

Do you think there is something to fix ?

File: recipes/libcurl/all/conanfile.py [hunk 19]
```diff
@@ -195,15 +186,19 @@ def requirements(self):
         if self.options.with_c_ares:
             self.requires("c-ares/1.18.1")
 
+    def package_id(self):
+        del self.info.settings.compiler
```
Comment (SpaceIm) [inline recipes/libcurl/all/conanfile.py:190]: what is that?

File: recipes/libcurl/all/conanfile.py [hunk 20]
```diff
@@ -344,111 +364,143 @@ def _patch_cmake(self):
...
-                params.append("--disable-ntlm")
+                tc.configure_args.append("--disable-ntlm")
 
         if not self.options.with_ntlm_wb:
-            params.append("--disable-ntlm-wb")
+            tc.configure_args.append("--disable-ntlm-wb")
 
-        if self.options.with_ca_bundle == False:
-            params.append("--without-ca-bundle")
+        if self.options.with_ca_bundle is False:
```
Comment (SpaceIm) [inline recipes/libcurl/all/conanfile.py:459]: default value is `None`and `None is False` return `False`, not `True`

File: recipes/libcurl/all/conanfile.py (hunk 19 above)
Comment (MartinDelille) [reply recipes/libcurl/all/conanfile.py:190]: I didn't manage to find why I pushed this commit.

**Pull request merged**